src/ApiCore/ResourceTemplate/Parser.php

```diff
@@ -0,0 +1,206 @@
+            }
+            return new Segment(Segment::LITERAL_SEGMENT, null, $segmentString);
+        }
+    }
+
+    private static function parseVariableSegment($segmentStringWithoutBraces)
+    {
+        // Validate there are no nested braces
+        $nestedOpenBracket = strpos($segmentStringWithoutBraces, '{');
+        if ($nestedOpenBracket !== false) {
+            throw new ValidationException(
+                "Unexpected '{' parsing segment $segmentStringWithoutBraces at index $nestedOpenBracket"
+            );
+        }
+
+        $equalsIndex = strpos($segmentStringWithoutBraces, '=');
+        if ($equalsIndex === false) {
+            // If the variable does not contain '=', we assume the pattern is '*' as per google.rpc.Http
+            $variableKey = $segmentStringWithoutBraces;
+            $nestedResource = new RelativeResourceTemplate("*");
+        } else {
+            $variableKey = substr($segmentStringWithoutBraces, 0, $equalsIndex);
+            $nestedResourceString = substr($segmentStringWithoutBraces, $equalsIndex + 1);
+            $nestedResource = new RelativeResourceTemplate($nestedResourceString);
+        }
+
+        if (!self::isValidLiteral($variableKey)) {
+            throw new ValidationException(
+                "Unexpected characters in variable name $variableKey"
+            );
+        }
+        return new Segment(Segment::VARIABLE_SEGMENT, $variableKey, null, $nestedResource);
+    }
+
+    /**
+     * @param string $literal
+     * @param string $path
+     * @param int $index
+     * @return string
+     * @throws ValidationException
+     */
+    private static function parseLiteralFromPath($literal, $path, &$index)
+    {
+        if (strlen($path) < ($index + strlen($literal))) {
+            throw self::parseError($literal, $path, $index);
+        }
+        $consumedLiteral = substr($path, $index, strlen($literal));
+        if ($consumedLiteral !== $literal) {
+            throw self::parseError($literal, $path, $index);
+        }
+        $index += strlen($literal);
+        return $consumedLiteral;
+    }
+
+    private static function parseError($literal, $path, $index)
+    {
+        return new ValidationException(
+            "Error parsing '$path' as index $index: " .
+            "expected '$literal'"
+        );
+    }
+
+    /**
+     * Check if $literal is a valid segment literal. Segment literals may only contain numbers,
+     * letters, and any of the following: .-~_
+     *
+     * @param string $literal
+     * @return bool
+     */
+    public static function isValidLiteral($literal)
+    {
+        return preg_match("/^[0-9a-zA-Z\\.\\-~_]+$/", $literal) === 1;
+    }
+
+    /**
+     * Check if $binding is a valid segment binding. Segment bindings may contain any characters
+     * except a forward slash ('/'), and may not be empty.
+     *
+     * @param $binding
+     * @return bool
+     */
+    public static function isValidBinding($binding)
```

Comment:
Could we group the public methods together? It just makes it nice and easy to scan the class and get an idea of what is exposed.

Reply:
Fixed in https://github.com/googleapis/gax-php/pull/209